### DISCUSSION TITLE
Interface Registry

### DISCUSSION THREAD

**Discussion by robert-zaremba** [near/NEPs#154] [Ideas]
For composability, smart contracts will need to have a priory knowledge about other smart contract interfaces. It will be extremely useful to store that information on chain:
+ smart contract designers could verify which interfaces are popular
+ when a smart contract will like to add / register other smart contract it may use the interface register to check for expected interfaces.

The mechanism is simple. A smart contract must call the registry and register itself. So only smart contract can do self registration, no one else can register a smart contract.
In the future version, however, we may consider a DAO mechanism, to allow the registration of older contract.

```rust
trait InterfaceRegistry {
    fn ir_implements(&self, interface: uint32, smart_contract: string) -> bool;
    fn ir_implements_many(&self, interface: uint32, smart_contract: Vec<string>) -> Vec<bool>;

    // smart contract self registration and un-registration
    fn ir_register(&mut self, interface: uint32);
    fn ir_unregister(&mut self, interface: uint32);
}
```

**Comment by robert-zaremba** [comment]
This can be implemented as a Curated List.

**Reply by mfornet** [reply]
@robert-zaremba who will be able to call `ir_register`? It could be either the contract itself, or a centralized authority. I wrote a similar approach some time ago: https://github.com/near/NEPs/discussions/275. I like the simplicity of this approach.

**Reply by robert-zaremba** [reply]
Probably yes, but then we need to have some authorization mechanism.

So, only smart contract should register and de-register itself. I will update the spec.

**Comment by mfornet** [comment]
Do we need to have `ir_unregister`? Some scenarios come to my mind: 
- Error interacting with registry (maybe impossible if access is properly set)
- Target contract changed (it was upgraded)

**Comment by frol** [comment]
I want to bring up the fact that [ABI for contracts on NEAR](https://github.com/near/abi) is approaching stabilization. It will embed the schema definition right in the contract itself. Yet, legacy contracts and JS / AS contracts are not supported yet.

**Reply by robert-zaremba** [reply]
will other smart contracts be able to query for the supported methods ?

**Reply by frol** [reply]
Absolutely, the ABI is exposed as a view-function call, so other contracts can just do a cross-contract call to fetch it.
